# Test Set Evaluation of Top Performing Models

## Libraries

In [1]:
import numpy as np
import matplotlib.pyplot as plt   
%matplotlib inline
from music21 import *
from IPython.display import Audio
from intervaltree import Interval,IntervalTree
import pandas as pd
import seaborn as sns

from sklearn.pipeline import make_pipeline
from sklearn.svm import SVC
import sklearn.model_selection as model_selection
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier

from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score
from sklearn.metrics import classification_report
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import confusion_matrix
from sklearn.decomposition import PCA
import tensorflow as tf

from collections import Counter
from imblearn.over_sampling import SMOTE
from sklearn.model_selection import train_test_split
import warnings
warnings.simplefilter(action='ignore')

### Import csv features which were extracted from earlier Data PreProcessing notebook

In [2]:
X_train_w = pd.read_csv('../data/X_train_w.csv',index_col = 0)
X_train_m = pd.read_csv('../data/X_train_m.csv',index_col = 0)
X_train_c = pd.read_csv('../data/X_train_c.csv',index_col = 0)
X_test_w = pd.read_csv('../data/X_test_w.csv',index_col = 0)
X_test_m = pd.read_csv('../data/X_test_m.csv',index_col = 0)
X_test_c = pd.read_csv('../data/X_test_c.csv',index_col = 0)
y_train_w = pd.read_csv('../data/y_train_w.csv',index_col = 0)
y_train_m = pd.read_csv('../data/y_train_m.csv',index_col = 0)
y_train_c = pd.read_csv('../data/y_train_c.csv',index_col = 0)
y_test_w = pd.read_csv('../data/y_test_w.csv',index_col = 0)
y_test_m = pd.read_csv('../data/y_test_m.csv',index_col = 0)
y_test_c = pd.read_csv('../data/y_test_c.csv',index_col = 0)

## Cross Validation Set Accuracy Results
| Model               | Feature Set | Model Variant                                           | Highest CV Score | Mean CV Score |
|---------------------|-------------|---------------------------------------------------------|------------------|---------------|
| Logistic Regression | Spectral    | Baseline                                                | 94.23%           | 89.75%        |
| Logistic Regression | Spectral    | PCA                                                     | 71.15%           | 68.83%        |
| Logistic Regression | Midi        | Baseline                                                | 76.92%           | 71.10%        |
| Logistic Regression | Midi        | PCA                                                     | 48.08%           | 41.49%        |
| Logistic Regression | Combined    | Baseline                                                | 92.31%           | 85.56%        |
| Logistic Regression | Combined    | PCA                                                     | 48.08%           | 45.64%        |
| Naive Bayes         | Spectral    | Baseline                                                | 71.69%           | 68.41%        |
| Naive Bayes         | Midi        | Baseline                                                | 75.00%           | 59.00%        |
| Naive Bayes         | Combined    | Baseline                                                | 75.47%           | 73.00%        |
| Decision Tree       | Spectral    | Baseline                                                | 80.77%           | 75.68%        |
| Decision Tree       | Midi        | Baseline                                                | 86.79%           | 84.03%        |
| Decision Tree       | Combined    | Baseline                                                | 94.23%           | 84.83%        |
| Random Forest       | Spectral    | Baseline                                                | 86.54%           | 83.29%        |
| Random Forest       | Spectral    | w/SMOTE                                                 | 100%             | 99.56%        |
| Random Forest       | Spectral    | w/SMOTE +  Hyperparameter Tuning                        | ?                | ?             |
| Random Forest       | Midi        | Baseline                                                | 94.34%           | 89.75%        |
| Random Forest       | Midi        | w/SMOTE                                                 | 99.56%           | 99.34%        |
| Random Forest       | Midi        | w/SMOTE + Hyperparameter Tuning                         | ?                | ?             |
| Random Forest       | Combined    | Baseline                                                | 90.39%           | 86.72%        |
| Random Forest       | Combined    | w/SMOTE                                                 | 100%             | 99.29%        |
| Random Forest       | Combined    | w/SMOTE + Hyperparameter Tuning                         | ?                | ?             |
| SVM                 | Spectral    | Baseline                                                | 67.39%           | 64.97%        |
| SVM                 | Spectral    | w/Hyperparameter Tuning                                 | 95.65%           | 91.35%        |
| SVM                 | Spectral    | w/Hyperparamater Tuning + PCA                           | 94.38%           | 90.47%        |
| SVM                 | Spectral    | w/Hyperparameter Tuning + PCA + SMOTE                   | 100%             | 99.38%        |
| SVM                 | Spectral    | w/Hyperparameter Tuning +  PCA + SMOTE + MinMaxScaler   | 77.33%           | 72.98%        |
| SVM                 | Spectral    | w/Hyperparameter Tuning +  PCA + SMOTE + StandardScaler | 98.67%           | 97.24%        |
| SVM                 | Midi        | Baseline                                                | 73.91%           | 73.17%        |
| SVM                 | Midi        | w/Hyperparameter Tuning                                 | ?                | ?             |
| SVM                 | Midi        | w/SMOTE                                                 | ?                | ?             |
| SVM                 | Midi        | w/PCA                                                   | ?                | ?             |
| SVM                 | Combined    | Baseline                                                | 73.91%           | 73.17%        |
| SVM                 | Combined    | w/Hyperparameter Tuning                                 | ?                | ?             |
| SVM                 | Combined    | w/SMOTE                                                 | ?                | ?             |
| SVM                 | Combined    | w/PCA                                                   | ?                | ?             |
| Neural Network      | Spectral    | Baseline (ADAM, 5 Hidden Layers)                        | 100%             | 100%          |
| Neural Network      | Midi        | Baseline (ADAM, 5 Hidden Layers)                        | 84.85%           | 83.55%        |
| Neural Network      | Combined    | Baseline (ADAM, 5 Hidden Layers)                        | 86.15%           | 81.82%        |

## Top Performing Models

| Model               | Feature Set | Model Variant                                            | Max CV Score | Mean CV Score |
|---------------------|-------------|----------------------------------------------------------|--------------|---------------|
| Neural Network      | Spectral    | Baseline (ADAM, 5 Hidden Layers)                         | 100%         | 100%          |
| Random Forest       | Spectral    | w/SMOTE                                                  | 100%         | 99.56%        |
| SVM                 | Spectral    | w/Hyperparameter   Tuning + PCA + SMOTE                  | 100%         | 99.38%        |
| Random Forest       | Midi        | w/SMOTE                                                  | 99.56%       | 99.34%        |
| Random Forest       | Combined    | w/SMOTE                                                  | 100%         | 99.29%        |
| SVM                 | Spectral    | w/Hyperparameter   Tuning + PCA + SMOTE + StandardScaler | 98.67%       | 97.24%        |
| SVM                 | Spectral    | w/Hyperparameter   Tuning                                | 95.65%       | 91.35%        |
| SVM                 | Spectral    | w/Hyperparamater   Tuning + PCA                          | 94.38%       | 90.47%        |
| Logistic Regression | Spectral    | Baseline                                                 | 94.23%       | 89.75%        |
| Random Forest       | Midi        | Baseline                                                 | 94.34%       | 89.75%        |

### Neural Network - Spectral - Baseline

In [14]:
model = tf.keras.Sequential([
    tf.keras.layers.Dense(176, activation='relu'),
    tf.keras.layers.Dense(100, activation='relu'),
    tf.keras.layers.Dense(90, activation='relu'),
    tf.keras.layers.Dense(80, activation='relu'),
    tf.keras.layers.Dense(70, activation='relu'),
    tf.keras.layers.Dense(60, activation='relu'),
    tf.keras.layers.Dense(21)
])
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
history = model.fit(X_train_w, y_train_w, epochs=200)
predictions = model.predict(X_test_w)
accuracy = accuracy_score(predictions,y_test_w)
print("Test Set Accuracy: {:.2%}".format(accuracy))

Epoch 1/200
8/8 [==============================] - 1s 4ms/step - loss: 5.0032 - accuracy: 0.3333 
Epoch 2/200
8/8 [==============================] - 0s 11ms/step - loss: 1.7950 - accuracy: 0.5152
Epoch 3/200
8/8 [==============================] - 0s 5ms/step - loss: 1.3109 - accuracy: 0.6104
Epoch 4/200
8/8 [==============================] - 0s 12ms/step - loss: 1.0787 - accuracy: 0.6364
Epoch 5/200
8/8 [==============================] - 0s 4ms/step - loss: 0.8014 - accuracy: 0.7662
Epoch 6/200
8/8 [==============================] - 0s 5ms/step - loss: 0.6730 - accuracy: 0.7965
Epoch 7/200
8/8 [==============================] - 0s 4ms/step - loss: 0.7845 - accuracy: 0.7792
Epoch 8/200
8/8 [==============================] - 0s 6ms/step - loss: 0.6179 - accuracy: 0.8398
Epoch 9/200
8/8 [==============================] - 0s 6ms/step - loss: 0.5584 - accuracy: 0.8442
Epoch 10/200
8/8 [==============================] - 0s 5ms/step - loss: 0.5389 - accuracy: 0.8225
Epoch 11/200
8/8 [========

8/8 [==============================] - 0s 6ms/step - loss: 0.0094 - accuracy: 1.0000
Epoch 85/200
8/8 [==============================] - 0s 5ms/step - loss: 0.0069 - accuracy: 1.0000
Epoch 86/200
8/8 [==============================] - 0s 14ms/step - loss: 0.0060 - accuracy: 1.0000
Epoch 87/200
8/8 [==============================] - 0s 4ms/step - loss: 0.0114 - accuracy: 0.9957
Epoch 88/200
8/8 [==============================] - 0s 5ms/step - loss: 0.0067 - accuracy: 1.0000
Epoch 89/200
8/8 [==============================] - 0s 4ms/step - loss: 0.0077 - accuracy: 1.0000
Epoch 90/200
8/8 [==============================] - 0s 4ms/step - loss: 0.0060 - accuracy: 1.0000
Epoch 91/200
8/8 [==============================] - 0s 4ms/step - loss: 0.0040 - accuracy: 1.0000
Epoch 92/200
8/8 [==============================] - 0s 4ms/step - loss: 0.0039 - accuracy: 1.0000
Epoch 93/200
8/8 [==============================] - 0s 3ms/step - loss: 0.0033 - accuracy: 1.0000
Epoch 94/200
8/8 [==============

8/8 [==============================] - 0s 3ms/step - loss: 2.2822e-04 - accuracy: 1.0000
Epoch 166/200
8/8 [==============================] - 0s 3ms/step - loss: 2.2662e-04 - accuracy: 1.0000
Epoch 167/200
8/8 [==============================] - 0s 2ms/step - loss: 2.1633e-04 - accuracy: 1.0000
Epoch 168/200
8/8 [==============================] - 0s 3ms/step - loss: 2.0816e-04 - accuracy: 1.0000
Epoch 169/200
8/8 [==============================] - 0s 5ms/step - loss: 2.0376e-04 - accuracy: 1.0000
Epoch 170/200
8/8 [==============================] - 0s 5ms/step - loss: 1.9798e-04 - accuracy: 1.0000
Epoch 171/200
8/8 [==============================] - 0s 3ms/step - loss: 1.9258e-04 - accuracy: 1.0000
Epoch 172/200
8/8 [==============================] - 0s 2ms/step - loss: 1.8797e-04 - accuracy: 1.0000
Epoch 173/200
8/8 [==============================] - 0s 5ms/step - loss: 1.8508e-04 - accuracy: 1.0000
Epoch 174/200
8/8 [==============================] - 0s 3ms/step - loss: 1.8165e-04 - a

ValueError: Classification metrics can't handle a mix of continuous-multioutput and multiclass targets

### Random Forest - Spectral - w/SMOTE

In [5]:
# todo: I think we need to re-import the data for this cell, if this is run multiple times then y_train_w gets overwritten
y_train_w = pd.read_csv('../data/y_train_w.csv',index_col = 0)

#Drop any ensemble types with counts below 6. This because Expected n_neighbors <= n_samples,  
#but n_samples = 3, n_neighbors = 6
counts = y_train_w['ensemble'].value_counts()
X_train_w = X_train_w[~y_train_w['ensemble'].isin(counts[counts < 6].index)]
y_train_w = y_train_w[~y_train_w['ensemble'].isin(counts[counts < 6].index)]

# Fit and apply the transform
X_SMOTE_w, y_SMOTE_w = SMOTE(random_state=101).fit_resample(X_train_w, y_train_w)

from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=1)
rf.fit(X_SMOTE_w, y_SMOTE_w)
predictions = rf.predict(X_test_w)
accuracy = accuracy_score(predictions,y_test_w)
print("Test Set Accuracy: {:.2%}".format(accuracy))

Test Set Accuracy: 75.76%


### SVM - Spectral - w/Hyperparameter Tuning + PCA + SMOTE

In [7]:
#Drop any ensemble types with counts below 6
counts = y_train_w['ensemble'].value_counts()
X_train_w_smote = X_train_w[~y_train_w['ensemble'].isin(counts[counts < 6].index)]
y_train_w_smote = y_train_w[~y_train_w['ensemble'].isin(counts[counts < 6].index)]

# fit and apply the transform
X_SMOTE, y_SMOTE = SMOTE().fit_resample(X_train_w_smote, y_train_w_smote.values.ravel())

#Perform PCA with 15 components
pca=PCA(n_components = 15)
pca.fit(X_SMOTE)
train_pca_w_smote = pca.transform(X_SMOTE)
X_test_w_pca = pca.transform(X_test_w)


from sklearn.model_selection import GridSearchCV
  
# defining parameter range
param_grid = {'C': np.logspace(-4, 4, 20), 
              'gamma': np.logspace(-3, 2, 6),
              'kernel': ['linear','rbf','poly']} 
  
grid = GridSearchCV(SVC(), param_grid , cv=5)
  
# fitting the model for grid search
grid.fit(train_pca_w_smote, y_SMOTE)

print(grid.best_params_)

#Create SVM model with best hyperparameters, PCA and SMOTE

svm = make_pipeline(SVC(kernel=grid.best_params_['kernel'], C = grid.best_params_['C'] , gamma=grid.best_params_['gamma']))
svm.fit(train_pca_w_smote, y_SMOTE)
predictions = svm.predict(X_test_w_pca)
accuracy = accuracy_score(predictions,y_test_w)
print("Test Set Accuracy: {:.2%}".format(accuracy))


{'C': 0.012742749857031334, 'gamma': 0.001, 'kernel': 'linear'}
Test Set Accuracy: 78.79%


### Random Forest - Midi - w/SMOTE

In [6]:
# todo: I think we need to re-import the data for this cell, if this is run multiple times then y_train_w gets overwritten
y_train_m = pd.read_csv('../data/y_train_m.csv',index_col = 0)

#Drop any ensemble types with counts below 6. This because Expected n_neighbors <= n_samples,  
#but n_samples = 3, n_neighbors = 6
counts = y_train_m['ensemble'].value_counts()
X_train_m = X_train_m[~y_train_m['ensemble'].isin(counts[counts < 6].index)]
y_train_m = y_train_m[~y_train_m['ensemble'].isin(counts[counts < 6].index)]

# Fit and apply the transform
X_SMOTE_m, y_SMOTE_m = SMOTE(random_state=101).fit_resample(X_train_m, y_train_m)

from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=1)
rf.fit(X_SMOTE_m, y_SMOTE_m)
predictions = rf.predict(X_test_m)
accuracy = accuracy_score(predictions,y_test_m)
print("Test Set Accuracy: {:.2%}".format(accuracy))


Test Set Accuracy: 83.33%


### Random Forest - Combined - w/SMOTE

In [8]:
# todo: I think we need to re-import the data for this cell, if this is run multiple times then y_train_w gets overwritten
y_train_c = pd.read_csv('../data/y_train_c.csv',index_col = 0)

#Drop any ensemble types with counts below 6. This because Expected n_neighbors <= n_samples,  
#but n_samples = 3, n_neighbors = 6
counts = y_train_c['ensemble'].value_counts()
X_train_c = X_train_c[~y_train_c['ensemble'].isin(counts[counts < 6].index)]
y_train_c = y_train_c[~y_train_c['ensemble'].isin(counts[counts < 6].index)]

X_SMOTE_c, y_SMOTE_c = SMOTE(random_state=101).fit_resample(X_train_c, y_train_c)

from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=1)
rf.fit(X_SMOTE_c, y_SMOTE_c)
predictions = rf.predict(X_test_c)
accuracy = accuracy_score(predictions,y_test_c)
print("Test Set Accuracy: {:.2%}".format(accuracy))

Test Set Accuracy: 80.30%


### SVM - Spectral - w/Hyperparameter Tuning + PCA + SMOTE + StandardScaler	

In [9]:
#### Standard Scaler Application

#Perform PCA with 15 components
pca=PCA(n_components = 15)
pca.fit(X_SMOTE)
train_pca_w_smote = pca.transform(X_SMOTE)
X_test_w_pca = pca.transform(X_test_w)


#Create SVM model with best hyperparameters, PCA and SMOTE and StandardScaler

svm = make_pipeline(StandardScaler() ,SVC(kernel=grid.best_params_['kernel'], C = grid.best_params_['C'] , gamma=grid.best_params_['gamma']))
svm.fit(train_pca_w_smote, y_SMOTE)


predictions = svm.predict(X_test_w_pca)
accuracy = accuracy_score(predictions,y_test_w)
print("Test Set Accuracy: {:.2%}".format(accuracy))


Test Set Accuracy: 74.24%


### SVM - Spectral - w/Hyperparameter Tuning

In [10]:
#Find best hyperparameters
param_grid = {'C': np.logspace(-4, 4, 20), 
              'gamma': np.logspace(-3, 2, 6),                     
              'kernel': ['linear','rbf','poly']} 
  
grid = GridSearchCV(SVC(), param_grid, cv=5)
  
# fitting the model for grid search
grid.fit(X_train_w, y_train_w.values.ravel())

print(grid.best_params_)


#Create Model with best hyperparameters

svm = make_pipeline(SVC(kernel=grid.best_params_['kernel'], C = grid.best_params_['C'] , gamma=grid.best_params_['gamma']))
svm.fit(X_train_w, y_train_w.values.ravel())
predictions = svm.predict(X_test_w)
accuracy = accuracy_score(predictions,y_test_w)
print("Test Set Accuracy: {:.2%}".format(accuracy))

{'C': 0.0018329807108324356, 'gamma': 0.001, 'kernel': 'linear'}
Test Set Accuracy: 77.27%


### SVM - Spectral - W/Hyperparameter Tuning + PCA

In [11]:
#Perform PCA with 15 components
pca=PCA(n_components = 15)
pca.fit(X_train_w)
train_pca_w = pca.transform(X_train_w)
X_test_w_pca = pca.transform(X_test_w)

from sklearn.model_selection import GridSearchCV
  
# defining parameter range
param_grid = {'C': np.logspace(-4, 4, 20), 
              'gamma': np.logspace(-3, 2, 6),
              'kernel': ['linear','rbf','poly']} 
  
grid = GridSearchCV(SVC(), param_grid , cv=5)
  
# fitting the model for grid search
grid.fit(train_pca_w, y_train_w.values.ravel())

print(grid.best_params_)

#Create Model with best hyperparameters and PCA
svm = make_pipeline(SVC(kernel=grid.best_params_['kernel'], C = grid.best_params_['C'] , gamma=grid.best_params_['gamma']))
svm.fit(train_pca_w, y_train_w.values.ravel())
predictions = svm.predict(X_test_w_pca)
accuracy = accuracy_score(predictions,y_test_w)
print("Test Set Accuracy: {:.2%}".format(accuracy))


{'C': 0.0018329807108324356, 'gamma': 0.001, 'kernel': 'linear'}
Test Set Accuracy: 74.24%


### Logistic Regression - Spectral - Baseline

In [12]:
#Create Baseline Logistic Model
lr = LogisticRegression(C = 0.5, solver='liblinear', multi_class='auto')
lr.fit(np.array(X_train_w), np.array(y_train_w))
predictions = lr.predict(X_test_w)
accuracy = accuracy_score(predictions,y_test_w)
print("Test Set Accuracy: {:.2%}".format(accuracy))


Test Set Accuracy: 78.79%


### Random Forest - Midi - Baseline

In [13]:
rf = RandomForestClassifier()
rf.fit(X_train_m, y_train_m)
predictions = rf.predict(X_test_m)
accuracy = accuracy_score(predictions,y_test_m)
print("Test Set Accuracy: {:.2%}".format(accuracy))

Test Set Accuracy: 83.33%


## Final Test Results

| Model               | Feature Set | Model Variant                                            | Mean CV Score | Test Score |
|---------------------|-------------|----------------------------------------------------------|---------------|------------|
| Neural Network      | Spectral    | Baseline (ADAM, 5 Hidden Layers)                         | 100%          | ?          |
| Random Forest       | Spectral    | w/SMOTE                                                  | 99.56%        | 75.76%     |
| SVM                 | Spectral    | w/Hyperparameter   Tuning + PCA + SMOTE                  | 99.38%        | 78.79%     |
| Random Forest       | Midi        | w/SMOTE                                                  | 99.34%        | 83.33%     |
| Random Forest       | Combined    | w/SMOTE                                                  | 99.29%        | 80.30%     |
| SVM                 | Spectral    | w/Hyperparameter   Tuning + PCA + SMOTE + StandardScaler | 97.24%        | 74.24%     |
| SVM                 | Spectral    | w/Hyperparameter   Tuning                                | 91.35%        | 77.27%     |
| SVM                 | Spectral    | w/Hyperparamater   Tuning + PCA                          | 90.47%        | 74.24%     |
| Logistic Regression | Spectral    | Baseline                                                 | 89.75%        | 78.79%     |
| Random Forest       | Midi        | Baseline                                                 | 89.75%        | 83.33%     |